core/src/main/kotlin/com/malinskiy/marathon/execution/Scheduler.kt

```diff
@@ -40,29 +40,44 @@ class Scheduler(
     private val track: Track,
     private val timer: Timer,
     private val testBundleIdentifier: TestBundleIdentifier?,
-    override val coroutineContext: CoroutineContext
-) : CoroutineScope {
+    context: CoroutineContext,
+    private val job: Job = Job()
+) : Actor<SchedulerMessage>(parent = null, context = context) {
 
-    private val job = Job()
     private val pools = ConcurrentHashMap<DevicePoolId, SendChannel<FromScheduler>>()
     private val poolingStrategy = configuration.poolingStrategy
 
     private val logger = MarathonLogging.logger("Scheduler")
 
     suspend fun execute() {
-        subscribeOnDevices(job)
         try {
+            subscribeOnDevices(job)
             withTimeout(deviceProvider.deviceInitializationTimeoutMillis) {
                 while (pools.isEmpty()) {
                     delay(100)
                 }
             }
+            /**
+             * this loop leads to that I couldn't set parent for Scheduler Actor
+             * So parent = null but we close Actor with finally block
+             */
+            for (child in job.children) {
+                child.join()
+            }
         } catch (e: TimeoutCancellationException) {
-            job.cancelAndJoin()
-            throw NoDevicesException("")
+            throw NoDevicesException("Timeout ${deviceProvider.deviceInitializationTimeoutMillis} ms reached while waiting devices initialization")
+        } finally {
+            close()
```

Comment:
Why we don't need to join the children now? Is that done as part of the close() call?